### DISCUSSION TITLE
Consider deriving from fedora-bootc

### DISCUSSION THREAD

**Discussion by cgwalters** [rhinstaller/anaconda#5888] [Ideas]
The Anaconda ISO has historically been built with lorax, and more recently osbuild.

One thing I'd love to see is support a spike on what it'd look like to build Anaconda as a bootc container image - and then the step of "create a live ISO" would be done via bootc-image-builder. I tried to create space for this in https://github.com/osbuild/bootc-image-builder/pull/165

The "internal" benefit here is basically having the center of gravity shift more towards container images - the disk images are secondary. Superficially, Anaconda from a user PoV would remain the same to start.

But this technical change would start to make more obvious the road to a few valuable things:

- As a customer who wants to create a bootc image with a custom kernel (for hardware enablement etc.) today precisely because the Anaconda ISO isn't built as a container but in an "osbuild way" inside bootc-image-builder, I can't literally just do `FROM <mycustom image> RUN dnf -y install anaconda` or so, but I'd also need to ensure that my custom kernel is provided as an RPM input to the bootc-image-builder process. The deep division between the two build systems is obvious here.
- It'd suddenly make it more obvious that we should support `podman run --privileged quay.io/fedora/anaconda ...` - and especially if we made it support [takeover installs](https://github.com/containers/bootc/pull/78) it'd be a quite flexible new way to do "reimaging" of systems, using Anaconda without dropping to rebooting into an ISO, etc. (Some of this may be possible of course today by just `dnf -y install anaconda`, I haven't tried, but running from a container is nicer I think for this)
- It'd be a great forcing function for us to generalize the case of "build a generic Live ISO" that we *definitely* want for a ton of use cases (as exists today for Fedora CoreOS but also the desktop Live ISOs etc.)

**Comment by jkonecny12** [comment]
I wonder, how would this work with our Dracut modules we need to have to make boot working. I'm not aware of Dracut modifications support for bootc. Is this possible?

**Reply by cgwalters** [reply]
https://docs.fedoraproject.org/en-US/bootc/initramfs/

**Comment by bam80** [comment]
> The Anaconda ISO has historically been built with lorax, and more recently osbuild.

Sorry could you give more info about the transition happened?
For what Lorax is still used, and for what (and since when) osbuild is?

I reported [a bug](https://github.com/weldr/lorax/issues/1451) for Lorax and now wondering what other projects are relevant, too?

**Reply by jkonecny12** [reply]
The whole situation is a chaotic, Lorax is still used, however it depends on the delivery target:
- Lorax is currently used for boot.iso (everywhere)
- Kiwi is newly used for some Live ISOs including spins on Fedora, however, the goal is to take over all boot.iso + Live ISO
- Image Builder is used mainly for bootc ISOs but also for on demand builds for RHEL

So as you can see, lorax is still used but a lot of tools are involved together which makes the situation quite chaotic and problematic. Also, we are not involved in any of the tools, so if you want a change, you need to contact them directly.

**Reply by cgwalters** [reply]
Did you see this bit in my initial comment?

> One thing I'd love to see is support a spike on what it'd look like to build Anaconda as a bootc container image - and then the step of "create a live ISO" would be done via bootc-image-builder. I tried to create space for this in https://github.com/osbuild/bootc-image-builder/pull/165

I have to emphasize the deep importance of the split between "container image" and "disk image". They are not the same thing. This proposal is about creating `quay.io/fedora/anaconda:rawhide` or so. The step of making an ISO from it should operate on that container image.

A different but important way to say this is that as someone who wants to e.g. test a patch to Anaconda, I can do so via `podman build` to make a *derived* container image `FROM quay.io/fedora/anaconda:rawhide`. Then I can pass it to a tool (such as bootc-image-builder) to materialize my custom container into an ISO.

In this process the tester/contributor can leverage their container (and bootc) knowledge and does not need to learn Kiwi XML or other things.

**Reply by cgwalters** [reply]
I filed https://issues.redhat.com/browse/RHEL-89177 as a downstream issue to try to better illustrate again what problem(s) I'm talking about.

**Comment by cgwalters** [comment]
@travier shared https://github.com/travier/fedora-kinoite/blob/main/fedora-kinoite-live/Containerfile

**Reply by p5** [reply]
To build on this, Universal Blue has started investing in Live ISOs built from our bootc images.  
We are pushing test ISOs with an in-house installer, though from looking at Timothée's repo, it seems very easy to switch over to anaconda once bootc is better supported.

Right now, we aren't bundling the installer inside the OCI, but rather splatting the OCI contents into a squashfs and dnf installing the installer (along with the other live ISO requirements) on that image.  We then push the OCI image into an overlay containers-storage inside that squashfs which bootc can install to-disk or to-filesystem from.

**Reply by jkonecny12** [reply]
Just for information, we are just now working on support for bootc kickstart command.

**Comment by cgwalters** [comment]
OK! I finally got around to doing some work here, see https://github.com/bootc-dev/bcvk/pull/202

I'd like to discuss upstreaming this container image build and get to the point where we ship `quay.io/fedora/anaconda:43` for example in the registry, and the ISO derives from it.

**Comment by cgwalters** [comment]
In an internal chat, https://github.com/osbuild/bootc-foundry/blob/main/rhel-10-installer was mentioned related to this. The status is it's not used for the pre-built ISOs yet, but would make sense to do.
